README.md
# docker-wordpress
A docker image to run wordpress

[![Docker Pulls](https://img.shields.io/docker/pulls/osixia/wordpress.svg)][hub]
[![Docker Stars](https://img.shields.io/docker/stars/osixia/wordpress.svg)][hub]

[hub]: https://hub.docker.com/r/osixia/wordpress/

Latest release: 5.4.1 - Wordpress 5.4.1 -  [Changelog](CHANGELOG.md) | [Docker Hub](https://hub.docker.com/r/osixia/wordpress) 

### Under the hood: osixia/web-baseimage

This image is based on osixia/web-baseimage.
More info: https://github.com/osixia/docker-web-baseimage

## Security
If you discover a security vulnerability within this docker image, please send an email to the Osixia! team at security@osixia.net. For minor vulnerabilities feel free to add an issue here on github.

Please include as many details as possible.

## Changelog

Please refer to: [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project follows Wordpress versioning.

## [5.4.1] - 2020-05-03
### Changed
  - Upgrade Wordpress version to 5.4.1

## [5.3.0] - 2019-11-15
### Changed
  - Upgrade Wordpress version to 5.3.0

## [5.2.4] - 2019-10-15
### Changed
  - Upgrade Wordpress version to 5.2.4

## [5.2.3] - 2019-09-29
## Added
  - Multiarch support

### Changed
  - Upgrade Wordpress version to 5.2.3
  - Upgrade baseimage to web-baseimage:1.2.0 (debian buster)

## [5.2.1] - 2019-05-29
### Changed
  - Upgrade Wordpress version to 5.2.1

## [5.1.1] - 2019-03-14
### Changed
  - Upgrade Wordpress version to 5.1.1

## [5.0.3] - 2019-01-30
### Changed
  - Upgrade Wordpress version to 5.0.3

## [4.9.7] - 2018-07-28
### Changed
  - Upgrade Wordpress version to 4.9.7

## [4.9.6] - 2018-05-24
### Changed
  - Upgrade Wordpress version to 4.9.6

## [4.9.5] - 2018-04-05
### Changed
  - Upgrade Wordpress version to 4.9.5

## [4.9.4] - 2018-02-07
### Changed
  - Upgrade Wordpress version to 4.9.4

## [4.9.2] - 2018-01-29
### Changed
  - Upgrade Wordpress version to 4.9.2

## [4.9.1] - 2017-12-04
### Changed
  - Upgrade Wordpress version to 4.9.1

## [4.9.0] - 2017-11-30
### Changed
  - Upgrade Wordpress version to 4.9

## [4.8.3] - 2017-11-16
### Changed
  - Upgrade Wordpress version to 4.8.3

## [4.8.2] - 2017-10-31
### Added
  - WORDPRESS_DISABLE_XMLRPC environment variable

### Changed
  - Upgrade Wordpress version to 4.8.2
  - Upgrade baseimage to web-baseimage:1.1.1

## [4.8.1] - 2017-09-20
### Added
  - opcache config
  - WORDPRESS_REMOVE_DEFAULT_THEMES and WORDPRESS_REMOVE_DEFAULT_PLUGINS environment variables
  - WORDPRESS_PRODUCTION environment variable

### Changed
  - Upgrade Wordpress version to 4.8.1
  - Upgrade baseimage to web-baseimage:1.1.0 (debian stretch, php7)
  - Optimise apache config

## [4.8.0] - 2017-06-13
### Changed
  - Upgrade Wordpress version to 4.8

## [4.7.5] - 2017-05-17
### Changed
  - Upgrade Wordpress version to 4.7.5

## [4.7.4] - 2017-04-25
### Changed
  - Upgrade Wordpress version to 4.7.4

## [4.7.3] - 2017-03-07
### Changed
  - Upgrade Wordpress version to 4.7.3

### Fixed
  - startup.sh wp-config.php link
  - .htaccess overwrite

## [4.7.2] - 2017-02-22
### Added
  - wp-content
  - ldap tls config

### Changed
  - Upgrade baseimage to web-baseimage:1.0.0
  - Upgrade Wordpress version to 4.7.2

## Versions before following the Wordpress versioning

## [0.1.2] - 2015-03-03
### Fixed
  - bootstrap config

## [0.1.1] - 2015-03-02
### Added
  - SSL certs directory
  - Bootstrap capabilities

## 0.1.0 - 2015-02-23
Initial release

[5.4.1]: https://github.com/osixia/docker-wordpress/compare/v5.3.0...v5.4.1
[5.3.0]: https://github.com/osixia/docker-wordpress/compare/v5.2.4...v5.3.0
[5.2.4]: https://github.com/osixia/docker-wordpress/compare/v5.2.3...v5.2.4
[5.2.3]: https://github.com/osixia/docker-wordpress/compare/v5.2.1...v5.2.3
[5.2.1]: https://github.com/osixia/docker-wordpress/compare/v5.1.1...v5.2.1
[5.1.1]: https://github.com/osixia/docker-wordpress/compare/v5.0.3...v5.1.1
[5.0.3]: https://github.com/osixia/docker-wordpress/compare/v4.9.7...v5.0.3
[4.9.7]: https://github.com/osixia/docker-wordpress/compare/v4.9.6...v4.9.7
[4.9.6]: https://github.com/osixia/docker-wordpress/compare/v4.9.5...v4.9.6
[4.9.5]: https://github.com/osixia/docker-wordpress/compare/v4.9.4...v4.9.5
[4.9.4]: https://github.com/osixia/docker-wordpress/compare/v4.9.2...v4.9.4
[4.9.2]: https://github.com/osixia/docker-wordpress/compare/v4.9.1...v4.9.2
[4.9.1]: https://github.com/osixia/docker-wordpress/compare/v4.9.0...v4.9.1
[4.9.0]: https://github.com/osixia/docker-wordpress/compare/v4.8.3...v4.9.0
[4.8.3]: https://github.com/osixia/docker-wordpress/compare/v4.8.2...v4.8.3
[4.8.2]: https://github.com/osixia/docker-wordpress/compare/v4.8.1...v4.8.2
[4.8.1]: https://github.com/osixia/docker-wordpress/compare/v4.8.0...v4.8.1
[4.8.0]: https://github.com/osixia/docker-wordpress/compare/v4.7.5...v4.8.0
[4.7.5]: https://github.com/osixia/docker-wordpress/compare/v4.7.4...v4.7.5
[4.7.4]: https://github.com/osixia/docker-wordpress/compare/v4.7.3...v4.7.4
[4.7.3]: https://github.com/osixia/docker-wordpress/compare/v4.7.2...v4.7.3
[4.7.2]: https://github.com/osixia/docker-wordpress/compare/v0.1.2...v4.7.2
[0.1.2]: https://github.com/osixia/docker-wordpress/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/osixia/docker-wordpress/compare/v0.1.0...v0.1.1


test/test_helper.bash
setup() {
  IMAGE_NAME="$NAME:$VERSION"
}

# function relative to the current container / image  
build_image() {
  #disable outputs
  docker build -t $IMAGE_NAME $BATS_TEST_DIRNAME/../image &> /dev/null
}

run_image() {
  CONTAINER_ID=$(docker run $@ -d $IMAGE_NAME)
  CONTAINER_IP=$(get_container_ip $CONTAINER_ID)
}

start_container() {
  start_container_by_cid $CONTAINER_ID
}

stop_container() {
  stop_container_by_cid $CONTAINER_ID
}

remove_container() {
 remove_container_by_cid $CONTAINER_ID
}

clear_container() {
  stop_container_by_cid $CONTAINER_ID
  remove_container_by_cid $CONTAINER_ID
}

is_service_running() {
  is_service_running_by_cid $CONTAINER_ID $1
}

wait_service() {
  wait_service_by_cid $CONTAINER_ID $@
}


# generic functions 
get_container_ip() {
   local IP=$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" $1)
   echo "$IP"
}

start_container_by_cid() {
  #disable outputs
  docker start $1 &> /dev/null
}

stop_container_by_cid() {
  #disable outputs
  docker stop $1 &> /dev/null
}

remove_container_by_cid() {
  #disable outputs
 docker rm $1 &> /dev/null
}

clear_container_by_cid() {
  for cid in "$@"
  do
    stop_container_by_cid $cid
    remove_container_by_cid $cid
  done
}

is_service_running_by_cid() {
  docker exec $1 ps cax | grep $2  > /dev/null
}

wait_service_by_cid() {

  cid=$1

  # first wait image init end
  while ! is_service_running syslog-ng
  do
    sleep 1
  done

  for service in "${@:2}"
  do
    # wait service
    while ! is_service_running_by_cid $cid $service
    do
      sleep 1
    done
  done

  sleep 5
}

image/service/wordpress/assets/wp-content/plugins/authldap/tests/AllTests.php
<?php
/**
 * $Id: AllTests.php 1047133 2014-12-17 14:55:04Z heiglandreas $
 *
 * authLdap - Authenticate Wordpress against an LDAP-Backend.
 * Copyright (c) 2008 Andreas Heigl<andreas@heigl.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * This file tests the basic LDAP-Tasks
 *
 * @category authLdap
 * @package authLdap
 * @subpackage tests
 * @author Andreas Heigl<andreas@heigl.org>
 * @copyright 2010 Andreas Heigl<andreas@heigl.org>
 * @license GPL
 * @since 21.09.2010
 */

require_once 'TestHelper.php';
require_once 'LdapTest.php';

if (!defined('PHPUnit_MAIN_METHOD')) {
    define('PHPUnit_MAIN_METHOD', 'AllTests::main');
}
/**
 * $Id: AllTests.php 1047133 2014-12-17 14:55:04Z heiglandreas $
 *
 * authLdap - Authenticate Wordpress against an LDAP-Backend.
 * Copyright (c) 2008 Andreas Heigl<andreas@heigl.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * This file tests the basic LDAP-Tasks
 *
 * @category authLdap
 * @package authLdap
 * @subpackage tests
 * @author Andreas Heigl<andreas@heigl.org>
 * @copyright 2010 Andreas Heigl<andreas@heigl.org>
 * @license GPL
 * @since 21.09.2010
 */
class AllTests
{
    public static function main()
    {
        $parameters = array();

        if (TESTS_GENERATE_REPORT && extension_loaded('xdebug')) {
            $parameters['reportDirectory'] = TESTS_GENERATE_REPORT_TARGET;
        }

        PHPUnit_TextUI_TestRunner::run(self::suite(), $parameters);
    }

    /**
     * Regular suite
     *
     * All tests except those that require output buffering.
     *
     * @return PHPUnit_Framework_TestSuite
     */
    public static function suite()
    {
        $suite = new PHPUnit_Framework_TestSuite('authLdap');

        $suite->addTestSuite('LdapTest');
        
        return $suite;
    }
}

if (PHPUnit_MAIN_METHOD == 'AllTests::main') {
    AllTests::main();
}


image/service/wordpress/assets/wp-content/plugins/authldap/tests/LdapTest.php
<?php
/**
 * $Id: LdapTest.php 1225557 2015-08-19 20:18:35Z heiglandreas $
 *
 * authLdap - Authenticate Wordpress against an LDAP-Backend.
 * Copyright (c) 2008 Andreas Heigl<andreas@heigl.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * This file tests the basic LDAP-Tasks
 *
 * @category authLdap
 * @package authLdap
 * @subpackage UnitTests
 * @author Andreas Heigl<andreas@heigl.org>
 * @copyright 2010 Andreas Heigl<andreas@heigl.org>
 * @license GPL
 * @since 21.09.2010
 */

/** ldap */
require_once 'ldap.php';

class LdapTest extends PHPUnit_Framework_TestCase
{
    /**
     *
     * @dataProvider dpInstantiateLdapClass
     * @param array $expected
     * @param array $given
     */
    public function testInstantiateLdapClass($expected, $given)
    {
        $ldap = new LDAP($expected[0], $expected[1]);
        foreach ($given as $key => $value) {
            $this -> assertAttributeEquals($value, '_' . $key, $ldap);
        }
    }

    /**
     * @dataProvider dpExceptionsWhenInstantiatingLdapClass
     * @expectedException Exception
     * @param array $expected
     */
    public function testExceptionsWhenInstantiatingLdapClass($expected)
    {
        new LDAP($expected);
    }

    public function dpInstantiateLdapClass()
    {
        return array (
            array (
             array ('ldap://uid=jondoe,cn=users,cn=example,c=org:secret@ldap.example.org/cn=example,c=org', true),
             array (
              'username' => 'uid=jondoe,cn=users,cn=example,c=org',
              'password' => 'secret',
              'server'   => 'ldap.example.org',
              'baseDn'   => 'cn=example,c=org',
              'debug'    => true
             )
            ),
            array (
             array ('ldap://uid=jondoe,cn=users,cn=example,c=org@ldap.example.org/cn=example,c=org', true),
             array (
              'username' => 'uid=jondoe,cn=users,cn=example,c=org',
              'password' => '',
              'server'   => 'ldap.example.org',
              'baseDn'   => 'cn=example,c=org',
              'debug'    => true
             )
            ),
            array(
             array ('ldap://ldap.example.org/cn=example,c=org', true),
             array (
              'username' => 'anonymous',
              'password' => '',
              'server'   => 'ldap.example.org',
              'baseDn'   => 'cn=example,c=org',
              'debug'    => true
             )
            ),
//            array(
//             array ('ldap://ldap.example.org', true),
//             array (
//              'username' => 'anonymous',
//              'password' => '',
//              'server'   => 'ldap.example.org',
//              'baseDn'   => '',
//              'debug'    => true
//             )
//            ),
            array(
             array ('ldap://uid=jondoe,cn=users,cn=example,c=org:secret@ldap.example.org/cn=example,c=org', false),
             array (
              'username' => 'uid=jondoe,cn=users,cn=example,c=org',
              'password' => 'secret',
              'server'   => 'ldap.example.org',
              'baseDn'   => 'cn=example,c=org',
              'debug'    => false
             )
            ),
        );
    }

    public function dpExceptionsWhenInstantiatingLdapClass()
    {
        return array (
                array('ldap://ldap.example.org'),
                array('ldap://foo:bar@/cn=example,c=org'),
                array('http://ldap.example.org'),
                array('fooBar'),
                array('ldap://ldap.example.org/'),
                array('()123üäö'),
               );
    }
}


image/service/wordpress/startup.sh
#!/bin/bash -e
# this script is run during the container start

# set -x (bash debug) if log level is trace
# https://github.com/osixia/docker-light-baseimage/blob/stable/image/tool/log-helper
log-helper level eq trace && set -x

FIRST_START_DONE="${CONTAINER_STATE_DIR}/docker-wordpress-first-start-done"

#
# HTTPS config
#
if [ "${WORDPRESS_HTTPS,,}" == "true" ]; then

  log-helper info "Set apache2 https config..."

  # generate a certificate and key if files don't exists
  # https://github.com/osixia/docker-light-baseimage/blob/stable/image/service-available/:ssl-tools/assets/tool/ssl-helper
  ssl-helper "${WORDPRESS_SSL_HELPER_PREFIX}" "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/certs/$WORDPRESS_SSL_CRT_FILENAME" "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/certs/$WORDPRESS_SSL_KEY_FILENAME" "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/certs/$WORDPRESS_SSL_CA_CRT_FILENAME"

  # add CA certificat config if CA cert exists
  if [ -e "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/certs/$WORDPRESS_SSL_CA_CRT_FILENAME" ]; then
    sed -i "s/#SSLCACertificateFile/SSLCACertificateFile/g" "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/https.conf"
  fi

  ln -sf "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/https.conf" /etc/apache2/sites-available/wordpress.conf
#
# HTTP config
#
else
  log-helper info "Set apache2 http config..."
  ln -sf "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/http.conf" /etc/apache2/sites-available/wordpress.conf
fi

if [ "${WORDPRESS_DISABLE_XMLRPC,,}" == "true" ]; then
  sed -i --follow-symlinks "s/#disablexmlrpc//g" /etc/apache2/sites-available/wordpress.conf
fi

a2ensite wordpress | log-helper debug

#
# Wordpress directory is empty, we use the bootstrap
#
if [ ! "$(ls -A -I lost+found /var/www/wordpress)" ]; then

  log-helper info "Bootstap Wordpress..."

  mkdir -p /var/www/wordpress
  cp -R /var/www/wordpress_bootstrap/* /var/www/wordpress

  if [ "${WORDPRESS_REMOVE_DEFAULT_THEMES}" = "true" ]; then
    rm -rf /var/www/wordpress/wp-content/themes/*
  fi

  if [ "${WORDPRESS_REMOVE_DEFAULT_PLUGINS}" = "true" ]; then
    rm -rf /var/www/wordpress/wp-content/plugins/*
  fi

fi

# remove bootstrap directory
rm -rf /var/www/wordpress_bootstrap

# install plugins and themes
cp --remove-destination -Rf "${CONTAINER_SERVICE_DIR}"/wordpress/assets/wp-content/. /var/www/wordpress/wp-content

# if there is no config
if [ ! -e "/var/www/wordpress/wp-config.php" ] && [ -e "${CONTAINER_SERVICE_DIR}/wordpress/assets/config/wp-config.php" ]; then

    log-helper debug "link ${CONTAINER_SERVICE_DIR}/wordpress/assets/config/wp-config.php to /var/www/wordpress/wp-config.php"
    ln -sf "${CONTAINER_SERVICE_DIR}/wordpress/assets/config/wp-config.php" /var/www/wordpress/wp-config.php

fi

# container first start
if [ ! -e "$FIRST_START_DONE" ]; then

  # Add .htaccess
  [ -e "/var/www/wordpress/.htaccess" ] || cp -f "${CONTAINER_SERVICE_DIR}/wordpress/assets/apache2/.htaccess" /var/www/wordpress/.htaccess

  # set new install default theme
  if [ -n "$WORDPRESS_DEFAULT_THEME" ]; then
    sed -i "s/define( 'WP_DEFAULT_THEME', '[^']*'/define( 'WP_DEFAULT_THEME', '${WORDPRESS_DEFAULT_THEME}'/g" /var/www/wordpress/wp-includes/default-constants.php
  fi

  cp -f "${CONTAINER_SERVICE_DIR}/wordpress/assets/php7.3-fpm/opcache.ini" /etc/php/7.3/fpm/conf.d/opcache.ini

  if [ "${WORDPRESS_PRODUCTION}" = "true" ]; then
    sed -i "s/;opcache.validate_timestamps/opcache.validate_timestamps/g" /etc/php/7.3/fpm/conf.d/opcache.ini
  fi

  touch "${FIRST_START_DONE}"
fi

# Fix file permission
find /var/www/ -type d -exec chmod 755 {} \;
find /var/www/ -type f -exec chmod 644 {} \;
chown www-data:www-data -R /var/www

# symlinks special (chown -R don't follow symlinks)
if [ -e "/var/www/wordpress/wp-config.php" ]; then

  chown www-data:www-data /var/www/wordpress/wp-config.php
  chmod 400 /var/www/wordpress/wp-config.php

fi


image/service/wordpress/container-start.sh
#!/bin/bash -e
# this script is run during the container start

FIRST_START_DONE="/etc/docker-wordpress-first-start-done"

# container first start
if [ ! -e "$FIRST_START_DONE" ]; then

  # create wordpress vhost
  if [ "${HTTPS,,}" == "true" ]; then

    # check certificat and key or create it
    /sbin/ssl-helper "/container/service/wordpress/assets/apache2/certs/$SSL_CRT_FILENAME" "/container/service/wordpress/assets/apache2/certs/$SSL_KEY_FILENAME"

    # add CA certificat config if CA cert exists
    if [ -e "/container/service/wordpress/assets/apache2/certs/$SSL_CA_CRT_FILENAME" ]; then
      sed -i "s/#SSLCACertificateFile/SSLCACertificateFile/g" /container/service/wordpress/assets/apache2/wordpress-ssl.conf
    fi

    a2ensite wordpress-ssl

  else
    a2ensite wordpress
  fi

  # wordpress directory is empty, we use the bootstrap
  if [ ! "$(ls -A /var/www/wordpress)" ]; then
    mkdir -p /var/www/wordpress/sources
    cp -R /var/www/wordpress_bootstrap/* /var/www/wordpress/sources
    rm -rf /var/www/wordpress_bootstrap

    # Move wp-config.php
    mv /var/www/wordpress/sources/wp-config-sample.php /var/www/wordpress/wp-config.php

    # set new install default theme
    if [ -n "$WORDPRESS_DEFAULT_THEME" ]; then
      sed -i "s/define( 'WP_DEFAULT_THEME', '[^']*'/define( 'WP_DEFAULT_THEME', '${WORDPRESS_DEFAULT_THEME}'/g" /var/www/wordpress/sources/wp-includes/default-constants.php
    fi

    # set db configuration
    sed -i "s/define('DB_NAME', '[^']*'/define('DB_NAME', '${DB_NAME}'/g" /var/www/wordpress/wp-config.php
    sed -i "s/define('DB_USER', '[^']*'/define('DB_USER', '${DB_USER}'/g" /var/www/wordpress/wp-config.php
    sed -i "s/define('DB_PASSWORD', '[^']*'/define('DB_PASSWORD', '${DB_PASSWORD}'/g" /var/www/wordpress/wp-config.php
    sed -i "s/define('DB_HOST', '[^']*'/define('DB_HOST', '${DB_HOST}'/g" /var/www/wordpress/wp-config.php
    sed -i "s/$table_prefix  = '[^']*'/$table_prefix  = '${DB_TABLE_PREFIX}'/g" /var/www/wordpress/wp-config.php

    # set authentication unique keys and salts.
    get_salt () {
      salt=$(</dev/urandom tr -dc '1324567890#<>,()*.^@$% =-_~;:|{}[]+!`azertyuiopqsdfghjklmwxcvbnAZERTYUIOPQSDFGHJKLMWXCVBN' | head -c64 | tr -d '\\')
    }

    toSalt=( AUTH_KEY SECURE_AUTH_KEY LOGGED_IN_KEY NONCE_KEY AUTH_SALT SECURE_AUTH_SALT LOGGED_IN_SALT NONCE_SALT )
    for key in ${toSalt[*]}
    do
      get_salt
      sed -i "s/define('${key}', *'[^']*'/define('${key}', '${salt}'/g" /var/www/wordpress/wp-config.php
    done

  fi

  touch $FIRST_START_DONE
fi

# Fix file permission
find /var/www/ -type d -exec chmod 755 {} \;
find /var/www/ -type f -exec chmod 644 {} \;
chmod 400 /var/www/wordpress/wp-config.php
chown www-data:www-data -R /var/www

exit 0


image/service/wordpress/assets/wp-content/plugins/authldap/ldap.php
<?php
/**
 * $Id: ldap.php 1526227 2016-11-01 19:21:26Z heiglandreas $
 *
 * authLdap - Authenticate Wordpress against an LDAP-Backend.
 * Copyright (c) 2008 Andreas Heigl<andreas@heigl.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * This file handles the basic LDAP-Tasks
 *
 * @author Andreas Heigl<andreas@heigl.org>
 * @package authLdap
 * @category authLdap
 * @since 2008
 */
namespace Org_Heigl\AuthLdap;

use Exception;

class LDAP
{
    private $_server = '';

    private $_scheme = 'ldap';

    private $_port = 389;

    private $_baseDn = '';

    private $_debug = false;
    /**
     * This property contains the connection handle to the ldap-server
     *
     * @var Ressource
     */
    private $_ch = null;

    private $_username = '';

    private $_password = '';

    private $_starttls = false;

    public function __construct($URI, $debug = false, $starttls = false)
    {
        $this->_debug=$debug;
        $array = parse_url($URI);
        if (! is_array($array)) {
            throw new Exception($URI . ' seems not to be a valid URI');
        }
        $url = array_map(function ($item) { return urldecode($item); }, $array);
        if (false === $url) {
            throw new Exception($URI . ' is an invalid URL');
        }
        if (! isset ( $url['scheme'] )) {
            throw new Exception($URI . ' does not provide a scheme');
        }
        if (0 !== strpos($url['scheme'], 'ldap')) {
            throw new Exception($URI . ' is an invalid LDAP-URI');
        }
        if (! isset ( $url['host'] )) {
            throw new Exception($URI . ' does not provide a server');
        }
        if (! isset ( $url['path'] )) {
            throw new Exception($URI . ' does not provide a search-base');
        }
        if (1 == strlen($url['path'])) {
            throw new Exception($URI . ' does not provide a valid search-base');
        }
        $this -> _server = $url['host'];
        $this -> _scheme = $url['scheme'];
        $this -> _baseDn = substr($url['path'], 1);
        if (isset ( $url['user'] )) {
            $this -> _username = $url['user'];
        }
        if ('' == trim($this -> _username)) {
            $this -> _username = 'anonymous';
        }
        if (isset ( $url['pass'] )) {
            $this -> _password = $url['pass'];
        }
        if (isset ( $url['port'] )) {
            $this -> _port = $url['port'];
        }
        $this->_starttls = $starttls;
    }

    /**
     * Connect to the given LDAP-Server
     *
     * @return LDAP
     * @throws AuthLdap_Exception
     */
    public function connect()
    {
        $this -> disconnect();
        if ('ldaps' == $this->_scheme && 389 == $this->_port) {
            $this->_port = 636;
        }

        $this->_ch = @ldap_connect($this->_scheme . '://' . $this->_server . ':' . $this -> _port);
        if (! $this->_ch) {
            throw new AuthLDAP_Exception('Could not connect to the server');
        }
        ldap_set_option($this->_ch, LDAP_OPT_PROTOCOL_VERSION, 3);
        ldap_set_option($this->_ch, LDAP_OPT_REFERRALS, 0);
        //if configured try to upgrade encryption to tls for ldap connections
        if ($this->_starttls) {
          ldap_start_tls($this->_ch);
        }
        return $this;
    }

    /**
     * Disconnect from a resource if one is available
     *
     * @return LDAP
     */
    public function disconnect()
    {
        if (is_resource($this->_ch)) {
            @ldap_unbind($this->_ch);
        }
        $this->_ch = null;
        return $this;
    }

    /**
     * Bind to an LDAP-Server with the given credentials
     *
     * @return LDAP
     * @throw AuthLdap_Exception
     */
    public function bind()
    {
        if (! $this->_ch) {
            $this->connect();
        }
        if (! is_resource($this->_ch)) {
            throw new AuthLDAP_Exception('No Resource-handle given');
        }
        $bind = false;
        if (( ( $this->_username )
            && ( $this->_username != 'anonymous') )
            && ( $this->_password != '' ) ) {
            $bind = @ldap_bind($this->_ch, $this->_username, $this->_password);
        } else {
            $bind = @ldap_bind($this->_ch);
        }
        if (! $bind) {
            throw new AuthLDAP_Exception('bind was not successfull: ' . ldap_error($this->_ch));
        }
        return $this;
    }

    public function getErrorNumber()
    {
        return @ldap_errno($this->_ch);
    }

    public function getErrorText()
    {
        return @ldap_error($this->_ch);
    }

    /**
     * This method does the actual ldap-serch.
     *
     * This is using the filter <var>$filter</var> for retrieving the attributes
     * <var>$attributes</var>
     *
     *
     * @param string $filter
     * @param array $attributes
     * @return array
     */
    public function search($filter, $attributes = array('uid'))
    {
        if (! is_Resource($this->_ch)) {
            throw new AuthLDAP_Exception('No resource handle avbailable');
        }
        $result = @ldap_search($this->_ch, $this->_baseDn, $filter, $attributes);
        if ($result === false) {
            throw new AuthLDAP_Exception('no result found');
        }
        $this->_info = @ldap_get_entries($this->_ch, $result);
        if ($this->_info === false) {
            throw new AuthLDAP_Exception('invalid results found');
        }
        return $this -> _info;
    }

    /**
     * This method sets debugging to ON
     */
    public function debugOn()
    {
        $this->_debug = true;
        return $this;
    }

    /**
     * This method sets debugging to OFF
     */
    public function debugOff()
    {
        $this->_debug = false;
        return $this;
    }

    /**
     * This method authenticates the user <var>$username</var> using the
     * password <var>$password</var>
     *
     * @param string $username
     * @param string $password
     * @param string $filter OPTIONAL This parameter defines the Filter to be used
     * when searchin for the username. This MUST contain the string '%s' which
     * will be replaced by the vaue given in <var>$username</var>
     * @return boolean true or false depending on successfull authentication or not
     */
    public function authenticate($username, $password, $filter = '(uid=%s)')
    {
        //return true;
        $this->connect();
        $this->bind();
        $res = $this->search(sprintf($filter, $username));
        if (! $res || ! is_array($res) || ( $res ['count'] != 1 )) {
            return false;
        }
        $dn = $res[0]['dn'];
        if ($username && $password) {
            if (@ldap_bind($this->_ch, $dn, $password)) {
                return true;
            }
        }
        return false;
    }
    /**
     * $this method loggs errors if debugging is set to ON
     */
    public function logError()
    {
        if ($this->_debug) {
            $_v = debug_backtrace();
            throw new AuthLDAP_Exception('[LDAP_ERROR]' . ldap_errno($this->_ch) . ':' . ldap_error($this->_ch), $_v[0]['line']);
        }
    }
}

class AuthLDAP_Exception extends Exception
{
    public function __construct($message, $line = null)
    {
        parent :: __construct($message);
        if ($line) {
            $this -> line = $line;
        }
    }
}
